src/qibosoq/qick_programs.py

```diff
@@ -0,0 +1,364 @@
+""" QickPrograms used by qibosoq to execute sequences and sweeps """
+
+from abc import ABC, abstractmethod
+from dataclasses import asdict
+from typing import List, Tuple
+
+import numpy as np
+from qibolab.instruments.rfsoc import QickProgramConfig
+from qibolab.platforms.abstract import Qubit
+from qibolab.pulses import Drag, Gaussian, Pulse, PulseSequence, PulseType, Rectangular
+from qibolab.sweeper import Parameter, Sweeper
+from qick import AveragerProgram, QickSoc, RAveragerProgram
+
+# conversion coefficients (in qibolab we use Hz and ns)
+HZ_TO_MHZ = 1e-6
+NS_TO_US = 1e-3
+
+
+class GeneralQickProgram(ABC):
+    """Abstract class for QickPrograms"""
+
+    def __init__(self, soc: QickSoc, qpcfg: QickProgramConfig, sequence: PulseSequence, qubits: List[Qubit]):
+        """In this function we define the most important settings.
+
+        In detail:
+            * max_gain, adc_trig_offset, max_sampling_rate, reps are imported from
+              qpcfg (runcard settings)
+            * relaxdelay (for each execution) is taken from qpcfg (runcard)
+            * syncdelay (for each measurement) is defined explicitly
+            * wait_initialize is defined explicitly
+            * super.__init__ (this will init AveragerProgram or RAveragerProgram)
+        """
+
+        self.soc = soc
+        self.soccfg = soc  # this is used by qick
+
+        self.sequence = sequence
+        self.qubits = qubits
+
+        # general settings
+        self.max_gain = qpcfg.max_gain
+        self.adc_trig_offset = qpcfg.adc_trig_offset
+        self.max_sampling_rate = qpcfg.sampling_rate
+        self.reps = qpcfg.reps
+
+        self.relax_delay = self.us2cycles(qpcfg.repetition_duration * NS_TO_US)
+        self.syncdelay = self.us2cycles(0)
+        self.wait_initialize = self.us2cycles(2.0)
+
+        self.pulses_registered = False
+
+        super().__init__(soc, asdict(qpcfg))
+
+    def declare_nqz_zones(self, sequence: PulseSequence):
+        """Declare nqz zone (1-2) for a given PulseSequence
+
+        Args:
+            sequence (PulseSequence): sequence of pulses to consider
+        """
+
+        ch_already_declared = []
+        for pulse in self.sequence:
+            if pulse.type is PulseType.DRIVE:
+                ch = self.qubits[pulse.qubit].drive.ports[0][1]
+            elif pulse.type is PulseType.READOUT:
+                ch = self.qubits[pulse.qubit].readout.ports[0][1]
```

Comment:
If you have a better idea of how `ports` should look, feel free to propose. 
The current layout `tuple(controller number, port number)` was inspired by QM, so I am not sure if it is optimal for other devices. Indexing `[0][1]` all the time does not look good but we should also keep in mind that it will be more complicated when you have more qubits and controllers.

Reply:
I think that the current implementation is ok... in a very distant future when the driver will control multiple board at the same time I will use both attributes somewhere :-)